### PR TITLE
SLING-9692: Add support for principal-based access control entries

### PR DESCRIPTION
**Pull request by karlpauls**

### CODE REVIEW

File: src/main/java/org/apache/sling/feature/cpconverter/accesscontrol/DefaultAclManager.java [hunk 1]
```diff
@@ -122,6 +134,8 @@ public void addRepoinitExtension(@NotNull List<VaultPackageAssembler> packageAss
                     .filter(entry -> getSystemUser(entry.getKey()).isPresent())
                     .map(Entry::getValue)
                     .flatMap(Collection::stream)
+                    // paths only should/need to be create with resource-based access control
+                    .filter(((Predicate<AccessControlEntry>) AccessControlEntry::isPrincipalBased).negate())
```
Comment (anchela) [inline src/main/java/org/apache/sling/feature/cpconverter/accesscontrol/DefaultAclManager.java:138]: if principal-based ac-setup is enforced for all service user permissions (and thus resource-based entries get converted to principal-based), no paths should be created either.... so in addition to filtering out principal-based entries, the whole block might be excluded if the enforce-prinicipal-based-ac-setup is turned on.

and we should have a test verifying this :)

File: src/main/java/org/apache/sling/feature/cpconverter/accesscontrol/DefaultAclManager.java [hunk 2]
```diff
@@ -166,32 +194,46 @@ private void addStatements(@NotNull SystemUser systemUser,
...
-                }
 
-                formatter.format("%n");
-            });
+        if (!principalEntries.isEmpty()) {
+            formatter.format("set principal ACL for %s%n", systemUser.getId());
+            principalEntries.forEach((entry, path) -> writeEntry(entry, path, formatter));
+            formatter.format("end%n");
+        }
+        if (!enforcePrincipalBased && !resourceEntries.isEmpty()) {
```
Comment (anchela) [inline src/main/java/org/apache/sling/feature/cpconverter/accesscontrol/DefaultAclManager.java:216]: isn't` !enforcePrincipalBased` redundant here?

File: src/main/java/org/apache/sling/feature/cpconverter/accesscontrol/DefaultAclManager.java [hunk 3]
```diff
@@ -166,32 +194,46 @@ private void addStatements(@NotNull SystemUser systemUser,
             return;
         }
 
-        Map<AccessControlEntry, String> entries = new LinkedHashMap<>();
+        Map<AccessControlEntry, String> resourceEntries = new LinkedHashMap<>();
+        Map<AccessControlEntry, String> principalEntries = new LinkedHashMap<>();
+
         authorizations.forEach(entry -> {
             String path = getRepoInitPath(entry.getRepositoryPath(), systemUser);
             if (path != null) {
```
Comment (anchela) [inline src/main/java/org/apache/sling/feature/cpconverter/accesscontrol/DefaultAclManager.java:202]: can/will `getRepoInitPath` ever return null? maybe it has the wrong nullability annotation?

File: src/test/java/org/apache/sling/feature/cpconverter/accesscontrol/EnforcePrincipalBasedTest.java [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+                "end" + System.lineSeparator();
+
+        String actual = repoinitExtension.getText();
+        assertEquals(expected, actual);
+
+        RepoInitParser repoInitParser = new RepoInitParserService();
+        List<Operation> operations = repoInitParser.parse(new StringReader(actual));
+        assertFalse(operations.isEmpty());
+    }
+}
```
Comment (anchela) [inline src/test/java/org/apache/sling/feature/cpconverter/accesscontrol/EnforcePrincipalBasedTest.java:163]: oh... seems to lack a new line

File: src/main/java/org/apache/sling/feature/cpconverter/accesscontrol/DefaultAclManager.java [hunk 5]
```diff
@@ -150,6 +164,20 @@ public void addRepoinitExtension(@NotNull List<VaultPackageAssembler> packageAss
...
+            } else {
+                String relpath = intermediatePath.toString().substring(parent.toString().length());
+                return supportedPrincipalBasedPath.toString() + relpath;
+            }
+        } else {
+            return intermediatePath.toString();
+        }
+    }
+
     private boolean aclStartsWith(RepoPath path) {
```
Comment (anchela) [inline src/main/java/org/apache/sling/feature/cpconverter/accesscontrol/DefaultAclManager.java:181]: this method is also called for groups, where we only wanted to create groups if a policy is located at the group-node (but not below).... the subsequent test for 'isBelow' will the cause the failure.... but IMO the correct statement here would to check if there is a acl whose access controlled path equals to the group path.
for users the starts-with is correct...

File: src/main/java/org/apache/sling/feature/cpconverter/accesscontrol/DefaultAclManager.java [hunk 6]
```diff
@@ -51,6 +51,9 @@
 
     private static final String CONTENT_XML_FILE_NAME = ".content.xml";
 
+    private final boolean enforcePrincipalBased;
+    private RepoPath supportedPrincipalBasedPath;
```
Comment (anchela) [inline src/main/java/org/apache/sling/feature/cpconverter/accesscontrol/DefaultAclManager.java:55]: may be final

File: src/main/java/org/apache/sling/feature/cpconverter/accesscontrol/DefaultAclManager.java [hunk 7]
```diff
@@ -150,6 +164,20 @@ public void addRepoinitExtension(@NotNull List<VaultPackageAssembler> packageAss
         }
     }
 
+    private String calculateIntermediatePath(RepoPath intermediatePath) {
```
Comment (anchela) [inline src/main/java/org/apache/sling/feature/cpconverter/accesscontrol/DefaultAclManager.java:167]: missing notnull

File: src/main/java/org/apache/sling/feature/cpconverter/accesscontrol/DefaultAclManager.java [hunk 8]
```diff
@@ -166,32 +194,46 @@ private void addStatements(@NotNull SystemUser systemUser,
...
+            formatter.format("end%n");
+        }
+        if (!enforcePrincipalBased && !resourceEntries.isEmpty()) {
+            formatter.format("set ACL for %s%n", systemUser.getId());
+            resourceEntries.forEach((entry, path) -> writeEntry(entry, path, formatter));
             formatter.format("end%n");
         }
     }
 
+    private void writeEntry(AccessControlEntry entry, String path, Formatter formatter) {
```
Comment (anchela) [inline src/main/java/org/apache/sling/feature/cpconverter/accesscontrol/DefaultAclManager.java:223]: missing notnull annotations

**Review by anchela — CHANGES_REQUESTED** [review]
initial findings from diff.... will look at the complete code next....

File: src/main/java/org/apache/sling/feature/cpconverter/accesscontrol/DefaultAclManager.java (hunk 2 above)
Comment (karlpauls) [reply src/main/java/org/apache/sling/feature/cpconverter/accesscontrol/DefaultAclManager.java:213]: True

File: src/main/java/org/apache/sling/feature/cpconverter/accesscontrol/DefaultAclManager.java (hunk 1 above)
Comment (karlpauls) [reply src/main/java/org/apache/sling/feature/cpconverter/accesscontrol/DefaultAclManager.java:138]: Yes, I think we can just wrap the complete block in an if statement.

File: src/main/java/org/apache/sling/feature/cpconverter/accesscontrol/DefaultAclManager.java (hunk 5 above)
Comment (karlpauls) [reply src/main/java/org/apache/sling/feature/cpconverter/accesscontrol/DefaultAclManager.java]: Right, that is why the group does both:
if (aclStartsWith(group.getPath())) {
                    formatter.format("create group %s with path %s%n", group.getId(), group.getIntermediatePath());
                }
                if (aclIsBelow(group.getPath())) {
                    throw new IllegalStateException("Detected policy on subpath of group: " + group);
                }
I think that is doing what you want - just in a roundabout way, no?

**Review by anchela — APPROVED** [review]

**Review by anchela — APPROVED** [review]

**Pull request merged**